File: pptx-service/renderers/causality.py
from builder import add_bg, add_textbox, add_rect, add_eyebrow, add_line
from theme import ThemeColors


def render(slide, data: dict, t: ThemeColors) -> None:
    add_bg(slide, t.bg)
    heading = data.get("heading", "")
    chain = data.get("chain", [])
    conclusion = data.get("conclusion", "")
    eyebrow = data.get("eyebrow", "")

    if eyebrow:
        add_eyebrow(slide, eyebrow, t)
    add_textbox(slide, 0.83, 0.5, 11.0, 1.0, heading,
                t.font_display, 44, t.text, bold=True)

    chain = chain[:5]
    n = len(chain)
    if n == 0:
        return

    node_w = 12.5 / n
    mid_y = 3.8

    for i, link in enumerate(chain):
        x = 0.42 + i * node_w
        cause = link.get("cause", "")
        because = link.get("because", "")
        is_last = i == n - 1

        add_rect(slide, x, mid_y - 0.6, node_w - 0.25, 1.2,
                 fill_hex=t.paper,
                 line_hex=t.accent if is_last else t.border,
                 line_width_pt=1.5 if is_last else 0.5)
        add_textbox(slide, x + 0.1, mid_y - 0.55, node_w - 0.45, 1.1,
                    cause, t.font_display, 22,
                    t.accent if is_last else t.text,
                    bold=True, align="center", wrap=True)

        if i < n - 1:
            ax = x + node_w - 0.25
            add_line(slide, ax, mid_y, ax + 0.25, mid_y, t.accent, 2.0)

        if because:
            add_textbox(slide, x, mid_y + 0.7, node_w - 0.2, 0.9,
                        because, t.font_body, 17, t.muted,
                        align="center", wrap=True)

    if conclusion:
        add_line(slide, 0.83, 6.5, 12.5, 6.5, t.border, 0.5)
        add_textbox(slide, 0.83, 6.6, 11.5, 0.55,
                    f"→ {conclusion}", t.font_body, 22, t.text, bold=True)

File: lib/pptx/slides/causality.ts
import type { CausalitySlide } from '@/lib/types';
import {
  px, py, SLIDE_W,
  setBackground, addRect, addLine, addTextBox, addHeading, addPageNumber,
  type PptxTheme,
} from '../helpers';
import type PptxGenJS from 'pptxgenjs';
type PSlide = ReturnType<InstanceType<typeof PptxGenJS>['addSlide']>;

export function render(slide: PSlide, data: CausalitySlide, t: PptxTheme, n: number, total: number): void {
  setBackground(slide, '#' + t.bg);
  addHeading(slide, data.heading, t, py(60));

  const chain = data.chain.slice(0, 5);
  const nodeW = px(260);
  const nodeH = py(180);
  const chainY = py(380);
  const totalW = SLIDE_W - 2 * t.paddingIn;
  const spacing = chain.length > 1 ? (totalW - chain.length * nodeW) / (chain.length - 1) : 0;

  chain.forEach((link, i) => {
    const x = t.paddingIn + i * (nodeW + spacing);
    addRect(slide, x, chainY, nodeW, nodeH, '#' + t.bg, { line: { color: t.accent, width: 2 } });
    addTextBox(slide, link.cause, {
      x: x + px(12), y: chainY + py(30), w: nodeW - px(24),
      fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, align: 'center', wrap: true,
    });

    if (i < chain.length - 1) {
      const arrowX = x + nodeW + spacing / 2;
      addTextBox(slide, '→', {
        x: arrowX - px(20), y: chainY + nodeH / 2 - py(20), w: px(40),
        fontSize: t.sectionSize, color: '#' + t.accent, fontFace: t.fontBody, align: 'center',
      });
      if (link.because) {
        addTextBox(slide, link.because, {
          x: x + nodeW + px(4), y: chainY + nodeH + py(10), w: spacing - px(8),
          fontSize: t.captionSize, color: '#' + t.muted, fontFace: t.fontBody, italic: true, align: 'center', wrap: true,
        });
      }
    }
  });

  if (data.conclusion) {
    addLine(slide, t.paddingIn, py(720), t.paddingIn, py(800), '#' + t.accent, 4);
    addTextBox(slide, data.conclusion, {
      x: t.paddingIn + px(20), y: py(725), w: SLIDE_W - 2 * t.paddingIn - px(20),
      fontSize: t.bodySize, color: '#' + t.text, fontFace: t.fontBody, bold: true, wrap: true,
    });
  }

  addPageNumber(slide, n, total, t);
}
